### PR TITLE
Add SkewNormal distribution 

### PR DESCRIPTION
**Pull request by azev77**
Add SkewNormal distribution based on: 
* [Skew normal distribution on Wikipedia](https://en.wikipedia.org/wiki/Skew_normal_distribution)
* [Discourse](https://discourse.julialang.org/t/skew-normal-distribution/21549/7)
* [SkewDist.jl](https://github.com/STOR-i/SkewDist.jl)

### CODE REVIEW

**Comment by johnczito** [comment]
Thanks! This would be a good add. My thoughts:

- I'm not sure about the fact that this implementation comes from a [discourse post](https://discourse.julialang.org/t/i-asked-for-help-optimizing-my-julia-code-now-its-someones-package/34679). At the very least, the source code must include much more explicit attribution.
- Needs tests. It would be good to include tests where you check against archived output from Azzalini's [`sn`](https://cran.r-project.org/web/packages/sn/index.html) package. Note that the unit tests for the univariates have a fairly standardized format that you can check out here: `Distributions.jl/test/ref/` and `Distributions.jl/test/univariates.jl`. If it's straightforward to ultimately cast the tests in that mold, that might be nice.
- Since this is a well-studied univariate, it should have a `cdf` method as soon as possible. In the (medium-term?) that would require that someone merge an implementation of [Owen's T function](https://en.wikipedia.org/wiki/Owen%27s_T_function) into `StatsFuns.jl`.

**Comment by azev77** [comment]
@johnczito thanks for responding. 
The PR is from wikipedia, @maximerischard's discourse post, @fairbrot's [Multivariate pkg](https://github.com/STOR-i/SkewDist.jl). 
1 I opened issue #1101 weeks before the PR, where I gave a shoutout to @maximerischard & @fairbrot
2 @fairbrot [encouraged this](https://github.com/STOR-i/SkewDist.jl/issues/3).
3 I haven't heard from @maximerischard. I emailed him on discourse.
4 I'm happy to add more explicit attribution
5 happy to add tests
Let me know anything else I can do.

**Comment by maximerischard** [comment]
Sorry I didn't see your discourse message. I'm very happy that you found my code snippet useful and for some of the code to find its way to `Distributions.jl`.

File: src/univariate/continuous/skewnormal.jl [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#### Sampling
+function rand(rng::AbstractRNG, d::SkewNormal)
+    u0 = randn(rng)
+    v = randn(rng)
+    δ = delta(d)
+    u1 = δ * u0 + √(1-δ^2) * v
+    return d.ξ + d.ω * sign(u0) * u1
+end
+
+## Fitting
```
Comment (matbesancon) [inline src/univariate/continuous/skewnormal.jl:62]: the commented lines should be removed, they seem to come from another file

**Comment by matbesancon** [comment]
this looks like a cool addition to Distributions. +1 to add tests

**Comment by azev77** [comment]
@johnczito @matbesancon @fairbrot @maximerischard 
I fine-tuned the code & added tests. Please have a look.

**Comment by maximerischard** [comment]
Looks great! I noticed there weren't any tests for `rand`, which also appears to be lacking for most distributions in the package. As a sanity check I compared a histogram of a million samples to the PDF, and it all looks good.

```julia
using PyPlot: PyPlot; plt=PyPlot
using Distributions
d1 = SkewNormal(1, 2, 3)
samples = rand(d1, 10^6);
plt.hist(samples; density=true, bins=1000)
let
    xx = range(-3.0, stop=8.0, length=1000)
    plt.plot(xx, pdf.(d1, xx))
end
```
index

I then also compared the empirical to the theoretical moments:
```julia
using Statistics
@show mean(samples), mean(d1)
@show std(samples), std(d1)
@show skewness(samples), skewness(d1)
@show kurtosis(samples), kurtosis(d1)
xx = range(-3.0, stop=8.0, length=100000)
@show xx[argmax(pdf.(d1, xx))], mode(d1)
```

```
(mean(samples), mean(d1)) = (2.5142092281272057, 2.513879513212096)
(std(samples), std(d1)) = (1.3088404926822659, 1.3069693261422424)
(skewness(samples), skewness(d1)) = (0.6675144945558559, 0.6670235701524081)
(kurtosis(samples), kurtosis(d1)) = (0.5017438999403141, 0.5097701294494131)
(xx[argmax(pdf.(d1, xx))], mode(d1)) = (1.946749467494675, 2.1058875860443687)
```

Again, all looking good except the mode is slightly off since it is an approximation (as is made clear in the comments in the code).

**Comment by johnczito** [comment]
> I noticed there weren't any tests for `rand`, which also appears to be lacking for most distributions in the package.

Yeah, it's a tricky one, as discussed in #143 and #1142. Anything would have some positive failure rate. I include some KS testing in the unit tests for the matrix variates, along with crude checks for the sample mean and covariance, but they're really just regression tests. They don't fail *for this seed.* If they start failing, it's just something to look into.

**Comment by azev77** [comment]
@johnczito @matbesancon 
Are we ready to roll?

**Comment by johnczito** [comment]
Looks good to me. Can you add tests checking that it agrees with `Normal` in the `α=0` case?

**Comment by azev77** [comment]
@johnczito @matbesancon 
Great idea. I added tests. 
Are we ready to roll?

**Comment by johnczito** [comment]
Sorry, for the sake of coverage, can you check every method that `Normal` and `SkewNormal` share? So `skewness`, `kurtosis`, `mgf`, and `cf` as well?

**Comment by azev77** [comment]
@johnczito done!
Ready to roll?

**Comment by azev77** [comment]
@johnczito please let me know if there is anything else I can do

**Comment by johnczito** [comment]
Looks good to me. I'm not sure what the [path forward](https://github.com/JuliaStats/StatsFuns.jl/issues/99) is on the cdf. Such as it is, this would be one of only two or three univariates that don't provide it, which means you can't really compute probabilities. I'm not sure how central folks think that functionality is.

**Comment by azev77** [comment]
@johnczito thanks for approving! 
1. there are several other properties I would like to add to SkewNormal in the future & I will think more about the cdf & possibly contributing to Owen's T
2. in mean time [SkewDist](https://github.com/STOR-i/SkewDist.jl/blob/master/src/skewnormal.jl).jl does 
`cdf(dist::SkewNormal, x::Real)=quadgk(t->pdf(dist,t), -Inf, x)[1]`
`quantile(dist::SkewNormal, β::Float64)=newton(x->cdf(dist,x) - β, dist.ξ)`
3. I'm new to PRs & Github, does this mean the current PR is approved & will be integrated in the package?

**Comment by johnczito** [comment]
@mschauer thoughts on merging this without a `cdf` method?

**Comment by azev77** [comment]
@mschauer do you think we can merge this PR?

File: src/univariate/continuous/skewnormal.jl [hunk 2]
```diff
@@ -0,0 +1,76 @@
+"""
+SkewNormal(ξ, ω, α)
+    The *skew normal distribution* is a continuous probability distribution
+    that generalises the normal distribution to allow for non-zero skewness.
+#
```
Comment (mschauer) [inline src/univariate/continuous/skewnormal.jl:5]: Remove
```
#
This code...
#
```

File: src/univariate/continuous/skewnormal.jl [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+#### Statistics
+delta(d::SkewNormal) = d.α/√(1+d.α^2)
+mean_z(d::SkewNormal) = √(2/π) * delta(d)
+std_z(d::SkewNormal) = 1 - 2/π * delta(d)^2
+
+mean(d::SkewNormal) = d.ξ + d.ω * mean_z(d)
+var(d::SkewNormal) = abs2(d.ω)*(1-mean_z(d)^2)
+std(d::SkewNormal) = √var(d)
+skewness(d::SkewNormal) = (4-π)/2 * mean_z(d)^3 / (1-mean_z(d)^2)^(3/2)
+kurtosis(d::SkewNormal) = 2*(π-3)*( (delta(d)* sqrt(2/π) )^4 / (1-2*(delta(d)^2)/π)^2)
```
Comment (mschauer) [inline src/univariate/continuous/skewnormal.jl:49]: Random whitespace

File: test/skewnormal.jl [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    #
+    d0 = SkewNormal(0.0, 1.0, 0.0)
+    @test SkewNormal() == d0
+    d3 = SkewNormal(0.5, 2.2, 0.0)
+    d4 = Normal(0.5, 2.2)
+    #
+    @test pdf(d3, 3.3) == Distributions.pdf(d4, 3.3)
+    @test pdf.(d3, 1:3) == Distributions.pdf.(d4, 1:3)
+    a = mean(d3), var(d3), std(d3)
+    #b = mean(d4), var(d4), std(d4)
```
Comment (mschauer) [inline test/skewnormal.jl:30]:
```suggestion
```

File: test/skewnormal.jl [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    b = Distributions.mean(d4), Distributions.var(d4), Distributions.std(d4)
+    @test  a == b
```
Comment (mschauer) [inline test/skewnormal.jl:32]:
```suggestion
    @test a == b
```

File: src/univariate/continuous/skewnormal.jl (hunk 2 above)
Comment (azev77) [reply src/univariate/continuous/skewnormal.jl]: done

**Comment by azev77** [comment]
@mschauer 
Thank you. Please have a look at the updated code.

**Comment by azev77** [comment]
@mschauer are we ready to roll?

**Comment by matbesancon** [comment]
seems travis is still failing

**Pull request merged**